### ISSUE TITLE
Add generics to ViewableData

### ISSUE THREAD

**Issue by Cambis** [silverstripeltd/silverstripe-framework-generics#19]
ViewableData causes some problems with typehinting in loops since it implements `IteratorAggregate`

**Comment by Cambis** [comment]
So two ways we could go about this

1. Make viewable data generic 

```php
/**
 * @template T of static
 * @implements \IteratorAggregate<array-key, T>
 */
 class ViewableData ....
 ```
 
 Then *every* subclass would need to
 ```php
 /**
  * @template T of static
  * @extends ViewableData<T>
  */
 class DataObject extends ViewableData ...
 ```
 
 This is kinda bad because we have a lot of DataObjects
 ```php
 /**
  * @extends DataObject<static>
  */
 class Member extends DataObject ...
 ```
 
2. Just remove `IteratorAggregate` from `ViewableData`? the source code makes a reference to the `control` block which I don't think is used anymore or deprecated? 🤔 
 
 This probably affects all classes that extend an `IteratorAggregate`
 

>  TLDR: If a class extends an `IteratorAggregate` then the parent class will need to have a generic type argument in order to get the proper typehinting in a foreach loop

**Comment by GuySartorelli** [comment]
Presumably we could just _not_ give generic types to non-lists that subclass `ViewableData` - so `DataObject` for example won't need it, because developers will almost never try to iterate over a dataobject record.

Ideally we'd remove iteration from `ViewableData` - but that'll have to be done in a major release. In the meantime, to avoid confusing your IDE, I think adding `@template T` to ViewableData, and then `@extends ViewableData<T>` only to _lists_ which subclass `ViewableData` would be okay.
